pkg/jobrunaggregator/jobrunaggregatorlib/util.go

```diff
@@ -0,0 +1,109 @@
+
+	var finishedJobRuns []jobrunaggregatorapi.JobRunInfo
+	var finishedJobRunNames []string
+	var unfinishedJobRuns []jobrunaggregatorapi.JobRunInfo
+	var unfinishedJobRunNames []string
+	for {
+		fmt.Println() // for prettier logs
+		// reset vars
+		finishedJobRuns = []jobrunaggregatorapi.JobRunInfo{}
+		unfinishedJobRuns = []jobrunaggregatorapi.JobRunInfo{}
+		finishedJobRunNames = []string{}
+		unfinishedJobRunNames = []string{}
+
+		if clock.Now().Before(readyAt) {
+			fmt.Printf("waiting until %v to collect more jobRuns before assessing finished or not. (now=%v)\n", readyAt, clock.Now())
+			time.Sleep(2 * time.Minute)
+			continue
+		}
+		fmt.Printf("it is %v, finished waiting until %v.\n", clock.Now(), readyAt)
+
+		relatedJobRuns, err := jobRunGetter.GetRelatedJobRuns(ctx)
+
+		if err != nil {
+			return finishedJobRuns, unfinishedJobRuns, finishedJobRunNames, unfinishedJobRunNames, err
+		}
+
+		if len(relatedJobRuns) == 0 {
+			return finishedJobRuns, unfinishedJobRuns, finishedJobRunNames, unfinishedJobRunNames, fmt.Errorf("found no related jobRuns")
+		}
+
+		for i := range relatedJobRuns {
+			jobRun := relatedJobRuns[i]
+			if !jobRun.IsFinished(ctx) {
+				fmt.Printf("%v/%v is not finished\n", jobRun.GetJobName(), jobRun.GetJobRunID())
+				unfinishedJobRunNames = append(unfinishedJobRunNames, jobRun.GetJobRunID())
+				unfinishedJobRuns = append(unfinishedJobRuns, jobRun)
+				continue
+			}
+
+			prowJob, err := jobRun.GetProwJob(ctx)
+			if err != nil {
+				fmt.Printf("  error reading prowjob %v: %v\n", jobRun.GetJobRunID(), err)
+				unfinishedJobRunNames = append(unfinishedJobRunNames, jobRun.GetJobRunID())
+				unfinishedJobRuns = append(unfinishedJobRuns, jobRun)
+				continue
+			}
+
+			if prowJob.Status.CompletionTime == nil {
+				fmt.Printf("%v/%v has no completion time for resourceVersion=%v\n", jobRun.GetJobName(), jobRun.GetJobRunID(), prowJob.ResourceVersion)
+				unfinishedJobRunNames = append(unfinishedJobRunNames, jobRun.GetJobRunID())
+				unfinishedJobRuns = append(unfinishedJobRuns, jobRun)
+				continue
+			}
+			finishedJobRuns = append(finishedJobRuns, jobRun)
+			finishedJobRunNames = append(finishedJobRunNames, jobRun.GetJobName()+jobRun.GetJobRunID())
+		}
+
+		summaryHTML := htmlForJobRuns(ctx, finishedJobRuns, unfinishedJobRuns)
+		if err := ioutil.WriteFile(filepath.Join(outputDir, "job-run-summary.html"), []byte(summaryHTML), 0644); err != nil {
+			return finishedJobRuns, unfinishedJobRuns, finishedJobRunNames, unfinishedJobRunNames, err
+		}
+
+		// ready or not, it's time to check
+		if clock.Now().After(timeToStopWaiting) {
+			fmt.Printf("waited long enough. Ready or not, here I come. (readyOrNot=%v now=%v)\n", timeToStopWaiting, clock.Now())
+			break
+		}
+
+		if len(unfinishedJobRunNames) > 0 {
+			fmt.Printf("found %d unfinished related jobRuns: %v\n", len(unfinishedJobRunNames), strings.Join(unfinishedJobRunNames, ", "))
+			time.Sleep(2 * time.Minute)
```

Comment:
This is probably a move (this screams "David wrote this fast"), but please use a construct like

```go
select{
case <- time.After():
    continue
case <- ctx.Done():
    return ctx.Error()
}
```

Reply:
Done